clean up acrobot examples.  implement AcrobotInput and AcrobotParams.…

RussTedrake opened the pull request:
…  add python bindings

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8109)
<!-- Reviewable:end -->

RussTedrake commented:
+@amcastro-tri for feature review, please

note that the line count is ~600, but it's mostly due to boilerplate json, parameter binding, and parameter binding test.  although I do admit that i felt like i opened pandora's box this morning. 

---

Review status: 0 of 31 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-:-L5naf6p1waauNeLNADD:bx98avo)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you @russtedrake, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Just minor comments.

---

Reviewed 33 of 33 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOHVD1xZe1RiIOSka:-L5pOHVD1xZe1RiIOSkb:boy4bb4) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L16)):*
> ```Python
>     Simulator
>     )
> ```

BTW blank line

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOl_nFOKUCTNL5XRB:-L5pOl_nFOKUCTNL5XRC:b2kf3g1) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L79)):*
> ```Python
>         initial_state = state.CopyToVector()
>         simulator.StepTo(1.0)
>         self.assertFalse((state.CopyToVector() == initial_state).any())
> ```

BTW, (just a suggestion) probably an interesting test would be to check if energy is conserved, for instance setting the damping parameters to zero.

---

*[examples/acrobot/acrobot_input.named_vector, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pDYgpskGFdT9rH3SA:-L5pDYgpskGFdT9rH3SB:bjruc5g) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_input.named_vector#L5)):*
> ```
> element {
>     name: "tau"
>     doc: "tau"
> ```

BTW in the other named vector files you have `units_doc` and default values.
Also, probably the `doc:` could be more verbose? like "torque at the elbow"? 

---

*[examples/acrobot/acrobot_plant.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5odf0FrP6gJ9Hm08Bs:-L5odf0FrP6gJ9Hm08Bt:b4ir1ty) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L23)):*
> ```cpp
> AcrobotPlant<T>::AcrobotPlant()
>     : systems::LeafSystem<T>(systems::SystemTypeTag<acrobot::AcrobotPlant>{}) {
>   this->DeclareVectorInputPort(AcrobotInput<T>());
> ```

BTW, wow, this declare's at construction look just beautiful, so clean.

---

*[examples/acrobot/acrobot_plant.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5p9nO11yldnZV5a8Lw:-L5p9nO11yldnZV5a8Lx:b-w0z6x1) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L42)):*
> ```cpp
>   p->set_m2(0.6178);
>   p->set_l1(0.2563);
>   p->set_l2(0);
> ```

is this correct? zero l2?

---

*[examples/acrobot/acrobot_plant.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5p9pdJVLeGRwyB-h0_:-L5p9pdJVLeGRwyB-h0a:bfhv15t) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L46)):*
> ```cpp
>   p->set_lc2(1.5651);
>   p->set_Ic1(-4.7443);
>   p->set_Ic2(-1.0068);
> ```

is this the result of the sys ID? negative inertias? 

---

*[examples/acrobot/acrobot_plant.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ocSHuWx7FZNLdC0_H:-L5ocSHuWx7FZNLdC0_I:bf1fg5o) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L31)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(AcrobotPlant)
> 
>   AcrobotPlant();
> ```

Is there a way to cross-reference the `AcrobotParams` here in the doc for the constructor? so that we can comment on what parameters define the model and their default values.

---

*[examples/acrobot/acrobot_plant.h, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5oXfRgKw3bgt9BDBiQ:-L5oXfRgKw3bgt9BDBiR:b-92podp) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L42)):*
> ```objc
> 
>   ///@{
>   /// Manipulator equation of Acrobot: M(q)q̈ + bias(q,q̇) = B*u.
> ```

BTW, didn't we agree on writing `C(q, v)v`?

---

*[examples/acrobot/acrobot_plant.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ob_8sAl6e6fZbV5ox:-L5ob_8sAl6e6fZbV5oy:bs4rn3r) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L46)):*
> ```objc
>   /// bias[2x1] includes the Coriolis term, gravity term and the damping term,
>   /// i.e.
>   /// bias[2x1] = C(q,v)*v - τ_g(q) + [b1*theta1;b2*theta2]
> ```

This entire doc doesn't quite render well in Doxygen. Maybe use a bulleted list?

---

*[examples/acrobot/acrobot_plant.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pCRKktzm3rd0OGscE:-L5pCRKktzm3rd0OGscF:b-xbbgh8) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L46)):*
> ```objc
>   /// bias[2x1] includes the Coriolis term, gravity term and the damping term,
>   /// i.e.
>   /// bias[2x1] = C(q,v)*v - τ_g(q) + [b1*theta1;b2*theta2]
> ```

BTW, damping should be on velocities. That is, `b₁θ̇₁` (www.unicodeit.net is just great!, have you tried it?)

---

*[examples/acrobot/acrobot_plant.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5obhEnC6pMGD-yp8GO:-L5obhEnC6pMGD-yp8GP:bdocirz) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L48)):*
> ```objc
>   /// bias[2x1] = C(q,v)*v - τ_g(q) + [b1*theta1;b2*theta2]
>   Vector2<T> dynamicsBiasTerm(const systems::Context<T>& context) const;
>   Matrix2<T> massMatrix(const systems::Context<T>& context) const;
> ```

These do not follow the GSG, capitalize. 

---

*[examples/acrobot/acrobot_plant.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5oc49wzeOaiW0s8qIh:-L5oc49wzeOaiW0s8qIi:b-vckzvy) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L53)):*
> ```objc
>   /// Evaluates the input port and returns the scalar value
>   /// of the commanded torque.
>   T get_tau(const systems::Context<T>& context) const {
> ```

BTW, return `const T&`?

---

*[examples/acrobot/acrobot_plant.h, line 74 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ocOiBfv96a_9xv_zQ:-L5ocOiBfv96a_9xv_zR:boy4bb4) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L74)):*
> ```objc
>     return get_mutable_state(&context->get_mutable_continuous_state());
>   }
> ```

BTW blank line

---

*[examples/acrobot/run_lqr_w_estimator.cc, line 116 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pEEVtdLDljCNxThWB:-L5pEEVtdLDljCNxThWC:b3yizb8) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/run_lqr_w_estimator.cc#L116)):*
> ```cpp
> 
>   // Set an initial condition near the upright fixed point.
>   auto& x0 = acrobot_w_encoder->get_mutable_acrobot_state(
> ```

BTW, probably clearer if we spell out the time here instead of using `auto`

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pLyHOTipQ1-tFR9Aq:-L5pLyHOTipQ1-tFR9Ar:b-wwakn5) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
> ```objc
>      * q̈ = M^-1 * (Bu - bias), where M = [a1,a2; a2,a3], B = [0;1], bias=[C0;C1]
>      * we have:
>      * q̈ = -C0*a2 + a3*(u-C1)
> ```

are you assuming `det(M)=1`?

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pNoNKz1p8wnG6AD8j:-L5pNoNKz1p8wnG6AD8k:b-8oqsd7) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
> ```objc
>      * q̈ = M^-1 * (Bu - bias), where M = [a1,a2; a2,a3], B = [0;1], bias=[C0;C1]
>      * we have:
>      * q̈ = -C0*a2 + a3*(u-C1)
> ```

q̈₂ instead of just q̈?

---

*[examples/acrobot/spong_controller.h, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pJtEu5KFwByHBfakE:-L5pJtEu5KFwByHBfakF:b2nozhu) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L130)):*
> ```objc
>       u = u_v(0, 0);
>     } else {
>       u = u_e + u_p;
> ```

BTW, u_p could've been computed here to avoid throwing away the previous computations.

---

*[examples/acrobot/spong_controller.h, line 148 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pKUpTwirgS0eprxg5:-L5pKUpTwirgS0eprxg6:bcbisoi) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L148)):*
> ```objc
>   const std::unique_ptr<systems::Context<T>> acrobot_context_;
>   Eigen::Matrix4d S_;
>   Eigen::Matrix<T, 1, 4> K_;
> ```

should this be on `<double>`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-:-L5pPki63X5NNRp7-Zpo:b-gm27su)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
thanks @amcastro-tri .  great review!
+@ggould-tri for platform tomorrow.

---

Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOHVD1xZe1RiIOSka:-L5qI4yZlnPl3_oigLbA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L16)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW blank line
</blockquote></details>

Done.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOl_nFOKUCTNL5XRB:-L5qI7eY8cjPAIg8tbQf:b-csy92g) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L79)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, (just a suggestion) probably an interesting test would be to check if energy is conserved, for instance setting the damping parameters to zero.
</blockquote></details>

i agree.  but the theme is to check correctness of the implementation (algorithm) in c++.  the python unit tests are really only checking that the spelling is correct, and the data flows correctly.  (and the energy methods aren't bound yet)

---

*[examples/acrobot/acrobot_input.named_vector, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pDYgpskGFdT9rH3SA:-L5qIaaOoizOu5jpVlLg:bc4sjck) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_input.named_vector#L5)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW in the other named vector files you have `units_doc` and default values.
Also, probably the `doc:` could be more verbose? like "torque at the elbow"? 
</blockquote></details>

Done.  Good call.

---

*[examples/acrobot/acrobot_plant.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5p9nO11yldnZV5a8Lw:-L5qIsyn4wUNOhgpMHWx:bl1348v) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L42)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is this correct? zero l2?
</blockquote></details>

it has no impact on the dynamics (so our sysid algorithm did not output a meaningful value).  i saw it, but don't have a better value to put in.  (it was there before, too)

---

*[examples/acrobot/acrobot_plant.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5p9pdJVLeGRwyB-h0_:-L5qJ7YxBtgc2qTdiqG9:bed6ejz) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is this the result of the sys ID? negative inertias? 
</blockquote></details>

yikes.  you're right.  i'll add a note (but can't actually fix the sysid right now).

---

*[examples/acrobot/acrobot_plant.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ocSHuWx7FZNLdC0_H:-L5qJqZ1D09JUICkIhN5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L31)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Is there a way to cross-reference the `AcrobotParams` here in the doc for the constructor? so that we can comment on what parameters define the model and their default values.
</blockquote></details>

Done.

---

*[examples/acrobot/acrobot_plant.h, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5oXfRgKw3bgt9BDBiQ:-L5qJtxm1ZmxcEHHi9Xr:b1nwhqe) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L42)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, didn't we agree on writing `C(q, v)v`?
</blockquote></details>

yes.  i changed as much as I could here to match our new conventions without changing the implementation.  calling it bias for now.  added a TODO

---

*[examples/acrobot/acrobot_plant.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ob_8sAl6e6fZbV5ox:-L5qKQVNp6MZu04nX0jf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This entire doc doesn't quite render well in Doxygen. Maybe use a bulleted list?
</blockquote></details>

Done.

---

*[examples/acrobot/acrobot_plant.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pCRKktzm3rd0OGscE:-L5qKYIseNIBzJJqP0Ky:bttcxs8) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, damping should be on velocities. That is, `b₁θ̇₁` (www.unicodeit.net is just great!, have you tried it?)
</blockquote></details>

good catch.  and yes, that's how i did all of the unicode you see here.

---

*[examples/acrobot/acrobot_plant.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5obhEnC6pMGD-yp8GO:-L5qKnd8u_M2wyw5Chda:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L48)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

These do not follow the GSG, capitalize. 
</blockquote></details>

Done.

---

*[examples/acrobot/acrobot_plant.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5oc49wzeOaiW0s8qIh:-L5qLQwTcGigcmeArIn0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L53)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, return `const T&`?
</blockquote></details>

Done.

---

*[examples/acrobot/acrobot_plant.h, line 74 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ocOiBfv96a_9xv_zQ:-L5qLTZZHNVIYPRGHvi6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L74)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW blank line
</blockquote></details>

Done.

---

*[examples/acrobot/run_lqr_w_estimator.cc, line 116 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pEEVtdLDljCNxThWB:-L5qLoEefn2KFZjDoa4T:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/run_lqr_w_estimator.cc#L116)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, probably clearer if we spell out the time here instead of using `auto`
</blockquote></details>

Done.

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pLyHOTipQ1-tFR9Aq:-L5qMHDF7FsWrcvumMIv:bow3yoe) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

are you assuming `det(M)=1`?
</blockquote></details>

there is a "strong inertial coupling" condition... yes.  it's in my notes.  
i'm actually going to clean up the math in my next pass (and move some of it to python).  

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pNoNKz1p8wnG6AD8j:-L5qMShtT9gepp279YtI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

q̈₂ instead of just q̈?
</blockquote></details>

Done.

---

*[examples/acrobot/spong_controller.h, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pJtEu5KFwByHBfakE:-L5qMzLDaPsUOYw47bWW:b-h7wvrv) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L130)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, u_p could've been computed here to avoid throwing away the previous computations.
</blockquote></details>

could call.  that's much cleaner.  i pulled a lot of the implementation down into the else.

---

*[examples/acrobot/spong_controller.h, line 148 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pKUpTwirgS0eprxg5:-L5qNDkXO6ecfzXAQ8oB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L148)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

should this be on `<double>`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-:-L5qO-jvL5oB8UiCeVSu:b-ytcfx0)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 16 of 24 files reviewed at latest revision, 13 unresolved discussions.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOHVD1xZe1RiIOSka:-L5qSK9FD7zhiCWvSAVB:bk06s3u) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L16)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

btw... the linter forced me to add it back.  python wants two blank lines before the class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 16 of 24 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOl_nFOKUCTNL5XRB:-L5qTTnHnjYwuv5ja1Yq:bb73djm) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L79)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i agree.  but the theme is to check correctness of the implementation (algorithm) in c++.  the python unit tests are really only checking that the spelling is correct, and the data flows correctly.  (and the energy methods aren't bound yet)
</blockquote></details>

SGTM

---

*[examples/acrobot/acrobot_plant.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5p9nO11yldnZV5a8Lw:-L5qU3-PCUVMAMMasscp:b-53hox1) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L42)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

it has no impact on the dynamics (so our sysid algorithm did not output a meaningful value).  i saw it, but don't have a better value to put in.  (it was there before, too)
</blockquote></details>

I see, only `lc2` matters. In that case, probably not even worth having it as part of the system's params? That would make it explicit this number doesn't appear in the equations.

---

*[examples/acrobot/acrobot_plant.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ocSHuWx7FZNLdC0_H:-L5qUKNPndePJunVEMef:biud1t3) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L31)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Excellent! Two comments:
- Comment with three bars for doxygen to generate doc.
- Does AcrobotParams get referenced in the doc even when it is an automatically generated class?

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pLyHOTipQ1-tFR9Aq:-L5qZGDOmpLzbwqhkvGD:b-ro14p3) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

there is a "strong inertial coupling" condition... yes.  it's in my notes.  
i'm actually going to clean up the math in my next pass (and move some of it to python).  
</blockquote></details>

Hum... I guess you could say that what you are doing is taking `q̈ = det(M)*y = det(M)*(- k_p * q₂ - k_d * q̇₂)`? sry, I just got to that conclusion by taking the inverse of M in my head, but I could be wrong. I can wait till you update these docs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 13 of 24 files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOl_nFOKUCTNL5XRB:-L5s3l8rb-zqdKe35UpN:b-8dzz6o) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L79)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

SGTM
</blockquote></details>

i went ahead and did it anyhow.  it will be nice to have the energy available in python.  ;-)

---

*[examples/acrobot/acrobot_plant.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5p9nO11yldnZV5a8Lw:-L5rxhdbkp7X5Gx2TWoh:b-sj6z5c) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.cc#L42)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I see, only `lc2` matters. In that case, probably not even worth having it as part of the system's params? That would make it explicit this number doesn't appear in the equations.
</blockquote></details>

It's true.  We've had it in there initially for visualization.  I will zap it.  

---

*[examples/acrobot/acrobot_plant.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5ocSHuWx7FZNLdC0_H:-L5s-j2wjB11HCFYnLpv:b-csf7dp) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/acrobot_plant.h#L31)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Excellent! Two comments:
- Comment with three bars for doxygen to generate doc.
- Does AcrobotParams get referenced in the doc even when it is an automatically generated class?
</blockquote></details>

sigh.  you're right.  Doxygen will not have it.  Will reference the file instead.

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pLyHOTipQ1-tFR9Aq:-L5s0FPCULpNhJYL5xz0:btm0pu8) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Hum... I guess you could say that what you are doing is taking `q̈ = det(M)*y = det(M)*(- k_p * q₂ - k_d * q̇₂)`? sry, I just got to that conclusion by taking the inverse of M in my head, but I could be wrong. I can wait till you update these docs.
</blockquote></details>

N.B. -- the old version of my notes on this is in section 3.5 of http://underactuated.mit.edu/underactuated.html?chapter=acrobot

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 22 of 33 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/examples/acrobot_py.cc, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5sjHvNHrlAVTVPnHj2:-L5sjHvOuCLHv3TogS1s:bp4ia67) ([raw file](https://github.com/robotlocomotion/drake/blob/6473e9210814b1fb9a657d1990084132e6a2820f/bindings/pydrake/examples/acrobot_py.cc#L29)):*
> ```cpp
>   // TODO(eric.cousineau): At present, we only bind doubles.
>   // In the future, we will bind more scalar types, and enable scalar
>   // conversion.
> ```

Rule-of-three (four, really) on this TODO.  Move it to an issue, cite the issue number instead?

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pLyHOTipQ1-tFR9Aq:-L5sl5WCF9tJGi_5Ir_H:boxlski) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

N.B. -- the old version of my notes on this is in section 3.5 of http://underactuated.mit.edu/underactuated.html?chapter=acrobot
</blockquote></details>

If this confuses Alejandro it will confuse other readers and should be in the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-:-L5slI1OkCQW23-i68_N:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/examples/acrobot_py.cc, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5sjHvNHrlAVTVPnHj2:-L5ssJSVEHMRvG3PeZHv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6473e9210814b1fb9a657d1990084132e6a2820f/bindings/pydrake/examples/acrobot_py.cc#L29)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Rule-of-three (four, really) on this TODO.  Move it to an issue, cite the issue number instead?
</blockquote></details>

Done.

---

*[examples/acrobot/spong_controller.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pLyHOTipQ1-tFR9Aq:-L5ssMLU305IrI21C0XT:b-s87yxb) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/examples/acrobot/spong_controller.h#L88)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

If this confuses Alejandro it will confuse other readers and should be in the comment.
</blockquote></details>

Added the link to the comment.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/examples/acrobot_py.cc, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5sjHvNHrlAVTVPnHj2:-L5suVwDTFESHpsws8-0:b4rse3p) ([raw file](https://github.com/robotlocomotion/drake/blob/6473e9210814b1fb9a657d1990084132e6a2820f/bindings/pydrake/examples/acrobot_py.cc#L29)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!
FYI:  Indentation is wrong.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 3 of 33 files at r1, 9 of 11 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/examples/test/acrobot_test.py, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5pOl_nFOKUCTNL5XRB:-L5t-XhazaaqtOFV89ZM:b-7s4s8d) ([raw file](https://github.com/robotlocomotion/drake/blob/46c08fb712983da344a41684ca78b5096094c345/bindings/pydrake/examples/test/acrobot_test.py#L79)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i went ahead and did it anyhow.  it will be nice to have the energy available in python.  ;-)
</blockquote></details>

cool!

---

*[examples/acrobot/spong_controller.h, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5t113nBByt3_JUJ6Tn:-L5t113nBByt3_JUJ6To:b-l1hot1) ([raw file](https://github.com/robotlocomotion/drake/blob/0b0bf430dd8bd5bf89cd4f7a177ab4fd3eea91cc/examples/acrobot/spong_controller.h#L103)):*
> ```objc
>        * We want q̈ = y = - k_p * q₂ - k_d * q̇₂
>        * Given Acrobot's manipulator equation:
>        * q̈ = M^-1 * (Bu - bias), where M = [a1,a2; a2,a3], B = [0;1],
> ```

Sry to be a drag, but I finally understood my confusion by looking at the implemented code. Where it says: `M = [a1,a2; a2,a3]` should instead say: `M⁻¹ = [a1,a2; a2,a3]`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 22 of 24 files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/examples/acrobot_py.cc, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5sjHvNHrlAVTVPnHj2:-L5ua-snYw1T5BDj5gq_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6473e9210814b1fb9a657d1990084132e6a2820f/bindings/pydrake/examples/acrobot_py.cc#L29)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Thanks!
FYI:  Indentation is wrong.
</blockquote></details>

Done.

---

*[examples/acrobot/spong_controller.h, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8109#-L5t113nBByt3_JUJ6Tn:-L5u_zTvryYh7b98sqh2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0b0bf430dd8bd5bf89cd4f7a177ab4fd3eea91cc/examples/acrobot/spong_controller.h#L103)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Sry to be a drag, but I finally understood my confusion by looking at the implemented code. Where it says: `M = [a1,a2; a2,a3]` should instead say: `M⁻¹ = [a1,a2; a2,a3]`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8109)*
<!-- Sent from Reviewable.io -->
